### PR TITLE
Stolon implementation.

### PR DESCRIPTION
**Pull request by hdevalence**
## TODO

- [ ] Split out minimal tor integration as described in #1643 
- [ ] Merge that PR, close this PR with a note

## Solution

This has a prototype implementation of parts of #1006. 

The actual transaction diffusion logic for Zebra is not implemented, but the `stolon` crate and the binary described in the RFC are, along with enabling changes to `zebra-network`.

There are still some rough edges to work out, but the `stolon` binary can successfully anonymously relay transactions to the Zcash network with no new supporting infrastructure:
```
~/c/z/z/stolon (stolon|✚1…) $ cat testnet-tx | cargo run -- -f 1 -p 127.0.0.1:9050
   Compiling zebra-network v3.0.0-alpha.0 (/home/hdevalence/code/zcash/zebra/zebra-network)
   Compiling stolon v0.1.0 (/home/hdevalence/code/zcash/zebra/stolon)
    Finished dev [unoptimized + debuginfo] target(s) in 5.60s
     Running `/home/hdevalence/code/zcash/zebra/target/debug/stolon -f 1 -p '127.0.0.1:9050'`
Sep 03 10:21:10.959 DEBUG stolon: opt=Opt { proxy_addr: 127.0.0.1:9050, testnet: false, fanout: 1 }
Sep 03 10:21:10.959 DEBUG stolon: using tor resolver to find peers seed_domain="mainnet.seeder.zfnd.org:8233"
Sep 03 10:21:14.264 DEBUG send_transaction{tor_addr=127.0.0.1:9050 peer_addr=46.4.192.189:8233}:connector{addr=0.0.0.0:8233}: zebra_network::peer::handshake: connecting to remote peer
Sep 03 10:21:14.264 DEBUG send_transaction{tor_addr=127.0.0.1:9050 peer_addr=46.4.192.189:8233}:connector{addr=0.0.0.0:8233}: zebra_network::peer::handshake: sending initial version message version=Version { version: Version(170012), services: (empty), timestamp: 2020-09-03T17:20:00Z, address_recv: (NODE_NETWORK, 0.0.0.0:8233), address_from: ((empty), 0.0.0.0:8233), nonce: Nonce(5736241258767978657), user_agent: "", start_height: Height(0), relay: false }
Sep 03 10:21:14.981 DEBUG send_transaction{tor_addr=127.0.0.1:9050 peer_addr=46.4.192.189:8233}:connector{addr=0.0.0.0:8233}: zebra_network::peer::handshake: got message from remote peer remote_msg=Version { version: Version(170011), services: NODE_NETWORK, timestamp: 2020-09-03T17:21:14Z, address_recv: (NODE_NETWORK, 195.144.21.219:56242), address_from: (NODE_NETWORK, 46.4.192.189:8233), nonce: Nonce(3866741022785186629), user_agent: "/MagicBean:3.0.0/", start_height: Height(959478), relay: true }
Sep 03 10:21:14.982 DEBUG send_transaction{tor_addr=127.0.0.1:9050 peer_addr=46.4.192.189:8233}:connector{addr=0.0.0.0:8233}: zebra_network::peer::handshake: got verack from remote peer
Sep 03 10:21:14.982 DEBUG send_transaction{tor_addr=127.0.0.1:9050 peer_addr=46.4.192.189:8233}:connector{addr=0.0.0.0:8233}: zebra_network::peer::handshake: constructing client, spawning server
Sep 03 10:21:14.984 DEBUG send_transaction{tor_addr=127.0.0.1:9050 peer_addr=46.4.192.189:8233}: stolon: sending transaction hash=TransactionHash("3956b54c11736f4ac5e2c474029cba8f5b83dca2e38f355337e20ce37fbdf064")
```
The tracing output above shows the fields filled into the handshake messages to avoid revealing distinguishers.

### CODE REVIEW

File: stolon/Cargo.toml
Comment (hdevalence) [inline stolon/Cargo.toml:15]: There are some rough spots in the `tokio-socks` crate (3rd party) but I think we could smooth them out and upstream some fixes.

File: stolon/src/lib.rs
Comment (hdevalence) [inline stolon/src/lib.rs]: This should maybe live in the (future) code that implements transaction relay, rather than here, because it may also be needed for the fallback relay logic.

File: stolon/src/lib.rs
Comment (hdevalence) [inline stolon/src/lib.rs]: `tor_resolve` returns a single IP, rather than multiple ones, and seems to do caching.  I'm not sure whether this is a `tokio-socks` limitation or a Tor limitation.  In any case, it prevents the fanout logic from actually working.

File: stolon/src/lib.rs
Comment (teor2345) [inline stolon/src/lib.rs]: Protocol Question:
Do seeders only supply Zcash instances on the default ports?

Style:
Maybe we should make this into a zebra-network function?
It seems like we've duplicated it a few times.

File: stolon/src/lib.rs
Comment (teor2345) [reply stolon/src/lib.rs]: **Analysis**

The Tor protocol allows multiple responses to a RESOLVE request:
https://gitweb.torproject.org/torspec.git/tree/tor-spec.txt#n1829

But Tor only sends a single IP address per RESOLVED cell:
https://github.com/torproject/tor/blob/master/src/feature/relay/dns.c#L491

And Tor caches DNS answers:
https://github.com/torproject/tor/blob/master/src/feature/relay/dns.c#L38

**Workaround**

As a workaround, make multiple resolve requests to different exits by modifying `tor_resolve` as follows:
```rust
- let sock = Self::execute_command(proxy, target, Authentication::None, Command::TorResolve).await?;
+ let sock = Self::execute_command(proxy, target, Authentication::Password { random_username, b"" }, Command::TorResolve).await?;
```
https://docs.rs/tokio-socks/0.3.0/src/tokio_socks/tcp.rs.html#99

Using a random username makes Tor use a different circuit for each request.

For details, see **IsolateSOCKSAuth** in: https://2019.www.torproject.org/docs/tor-manual.html.en#SocksPort

**Follow-Up**

`tokio-socks`

Using a new circuit (by supplying random authentication) should be supported via an argument to `tor_resolve`, and all of the tor functions.

Using tor's default slow exit rotation (10+ minutes) for resolves can impact your anonymity - depending on your particular threat model.

`tor`

tor could return multiple addresses in `RESOLVED` cells, saving load on the network.

File: stolon/src/lib.rs
Comment (hdevalence) [reply stolon/src/lib.rs]: Yeah, I'm not sure about this.  I don't know whether the seeders only supply Zcash instances that listen on the default ports, but on the other hand, because of the way that the DNS seeders work (DNS names can't resolve to a port), I don't see how we could learn the port except by guessing.  Maybe @gtank knows.

On the style question: yes, I think this probably makes sense to move into a function.  Maybe `Network::default_port(&self) -> u16`?

File: stolon/src/lib.rs
Comment (teor2345) [reply stolon/src/lib.rs]: **Alternate Workaround**

If you don't need to know the IP addresses of the Zcash nodes you're using, you can save time and network load by simply asking Tor to connect to the seeder hostname.

Then the Tor exit will resolve the seeder hostname for you, and connect to one of the nodes supplied by the seeder.

I just realised that this would work when reading this bitcoin ticket:
https://github.com/bitcoin/bitcoin/issues/6808

To get different nodes, you'll need to ask Tor for a separate circuit for each connection, by providing randomised authentication.

(Tor actually returns the remote IP address via the SOCKS5 protocol, but that information is not exposed by tokio-socks.)

Comment (teor2345) [reply stolon/src/lib.rs]: (You could also get a different circuit by using different seeder addresses, but since there are only 3 or 4 of those, depending on the network, that would limit your fanout.)

File: stolon/src/lib.rs
Comment (teor2345) [reply stolon/src/lib.rs]: > Yeah, I'm not sure about this. I don't know whether the seeders only supply Zcash instances that listen on the default ports, but on the other hand, because of the way that the DNS seeders work (DNS names can't resolve to a port), I don't see how we could learn the port except by guessing. Maybe @gtank knows.

I can't see any port filtering in [dnsseeder](https://github.com/zcashfoundation/dnsseeder) or [zcash-seeder](https://github.com/zcash/zcash-seeder).

Maybe it's not obvious, or maybe it was just left out?

> On the style question: yes, I think this probably makes sense to move into a function. Maybe `Network::default_port(&self) -> u16`?

Sounds good. Then we can replace the existing references to 8233 / 18233.

File: stolon/src/lib.rs
Comment (hdevalence) [reply stolon/src/lib.rs]: In the short term, I think your alternative fix (resolving the DNS name directly) seems like a good option.

Comment (hdevalence) [reply stolon/src/lib.rs]: Implemented in daad311324db72a7abf9f5481595e3887b5f5749

**Comment by hdevalence** [comment]
This PR includes supporting changes to the networking code. I sequenced the commits so that they're prior and independent, and I think it would be easier to review if I split them out into separate PRs.

**Comment by hdevalence** [comment]
Blocked on: #1013 #1014 #1015 #1016 , which contain the independent chunks of supporting work for this code. Once they're merged into main, I'll rebase this branch.

File: stolon/src/lib.rs
Comment (hdevalence) [inline stolon/src/lib.rs:79]: This uses `""` as the user-agent, because the user-agent should be a valid BIP14 string, and BIP14 strings are sequences of name-version pairs.  But we don't want to specify a version, because that's a distinguisher.  So we don't give a name, either.

File: stolon/src/lib.rs
Comment (teor2345) [inline stolon/src/lib.rs:39]: I just did a quick check to make sure this is true:
* [script](https://gist.github.com/teor2345/dbd45dd1bd372f06702106d8682650df)
* [list](https://gist.github.com/teor2345/a11fcd2e0a8ebe0fc0229cdd8a7517ea)

There are over 600 tor exits that support port 8233 - about half the number of exits:
https://metrics.torproject.org/relayflags.html

So there are enough exits to support this design. (And there are ~300 that support the testnet port, which is enough for testing.)

The largest relay has a weight that's around 1.5% of the total. So with a fanout of 3, the probability of choosing the same exit is at most 3%. (I derived an upper bound for the probability, using the weight of the largest exit. The exact probability is the sum of `p * (1 - p)^2` for all exits, which is bounded above by `(1 - p)^2` for the largest exit.)

I'm not sure if choosing the same Zcash node twice matters for privacy here. But if it does, we should adjust the fanout to compensate.

Edit: use the default fanout of 3 in the example

File: stolon/src/lib.rs
Comment (teor2345) [reply stolon/src/lib.rs]: Looks good!

File: stolon/src/lib.rs
Comment (teor2345) [inline stolon/src/lib.rs:42]: Should we choose between multiple DNS seeders at random?
Should we fallback to another seeder if the one we choose is down?

File: stolon/src/lib.rs
Comment (teor2345) [inline stolon/src/lib.rs:50]: Should we wrap this request in a timeout?
What should we do if the connect times out or fails?

File: stolon/src/lib.rs
Comment (teor2345) [inline stolon/src/lib.rs:69]: Should we wrap this request in a timeout?
What should we do if the connect times out or fails?

File: stolon/src/lib.rs
Comment (teor2345) [inline stolon/src/lib.rs:86]: Should we wrap this request in a timeout?
What should we do if the send times out or fails?

**Review by teor2345 — CHANGES_REQUESTED** [review]
I noticed a few places where we might want redundancy or timeouts.

What's the plan for testing this utility?

Will the testnet transaction in 74243a3b865b021f3bc923e4307c4815f79bac63 become part of an acceptance test?

File: stolon/src/lib.rs
Comment (hdevalence) [reply stolon/src/lib.rs:39]: I don't think that choosing the same Zcash node twice matters for privacy, because that node doesn't learn any new information about the client or the transaction.  The motivation for doing fanout is just to reduce the probability of failure in transaction diffusion.

File: stolon/src/lib.rs
Comment (hdevalence) [reply stolon/src/lib.rs:42]: If we could resolve DNS addresses to multiple IPs and collect the resulting sets (like we can do with https://doc.rust-lang.org/nightly/std/net/trait.ToSocketAddrs.html), I think this would be easier (we could assemble a list of candidates, and sample within that list). Unfortunately, because we do DNS resolution on a per-connection basis, we can't look up multiple seeders and combine results, so we'd have to pick one per connection attempt.

I'm not sure that we would be able to tell that an error occurred because the seeder is down, rather than for some other reason (e.g., transit through a hostile network that can and does modify data).

We could try to pick one domain per `send_transaction` call, e.g., selecting between the ZF seeder and the ECC seeder with 50%/50% probability.  I don't want to have a large list of DNS seeders, because I think it would be preferable to have a small number of well-maintained infrastructure components rather than a larger number of less well-maintained ones.

File: stolon/src/lib.rs
Comment (hdevalence) [reply stolon/src/lib.rs:50]: If the connection fails, the error is propagated upwards to the caller.  There shouldn't be a timeout here or in any other part of the body of the function, but `send_transaction` and `send_transaction_to` should have a comment like

https://github.com/ZcashFoundation/zebra/pull/1014/files#diff-f2a0a2df0cb5a189c9df6ab45e6c2c34R29-R31

```
///
/// Note that this method does not implement any timeout behavior, so callers may
/// want to layer it with a timeout as appropriate for their application.
///
```
and there should be a timeout applied in `main.rs`.

File: stolon/src/lib.rs
Comment (hdevalence) [reply stolon/src/lib.rs:69]: same rationale as here https://github.com/ZcashFoundation/zebra/pull/1008/files#r484533061

File: stolon/src/lib.rs
Comment (hdevalence) [reply stolon/src/lib.rs:86]: same rationale as https://github.com/ZcashFoundation/zebra/pull/1008/files#r484533061

File: stolon/src/lib.rs
Comment (teor2345) [reply stolon/src/lib.rs:39]: Ok, so it is very unlikely that we'll choose the same node 3 times, and that node is also failing.

It's also unlikely that we'll choose multiple nodes, and they are all failing.

File: stolon/src/lib.rs
Comment (teor2345) [reply stolon/src/lib.rs:42]: Yeah, Tor doesn't let you choose between multiple DNS addresses for the same connection.

If the goal is to reduce the probability of failure, then we should try to connect to each seeder once for every transaction. Then, if one seeder is down, the other seeder should handle the request.

For example, here is one possible fanout design:
1. Connect to the ZFND seeder
2. Connect to the ECC seeder
3. For the remaining fanouts, connect uniformly at random to either seeder

File: stolon/src/lib.rs
Comment (hdevalence) [reply stolon/src/lib.rs]: The `default_port` method is now present in `zebra-network`. But I think that this discussion point was subsumed by https://github.com/ZcashFoundation/zebra/pull/1008#discussion_r484156174

File: stolon/src/lib.rs
Comment (hdevalence) [reply stolon/src/lib.rs:42]: From an implementation point of view I think it would be preferable if all of the `send_transaction` tasks could be fully independent, but if we force choosing the ZF / ECC seeders at least once there would have to be some communication.  WDYT of the idea of connecting uniformly at random to either the ECC or ZF seeder in each `send_transaction` call?

**Comment by hdevalence** [comment]
Rebased onto main now that all of the supporting infrastructure changes are merged in.

I changed from "Draft PR" to "Ready for Review" because it is technically ready, but it's still a low priority relative to other work.  (The other supporting infrastructure changes like #1016 were higher priority, since we needed them for other reasons like #889).

**Comment by teor2345** [comment]
We're going to close this PR for now, because we're focused on NU5 validation.

Our follow up plan is in ticket #1643, we plan to work on it as part of `zebra-client`.

**Pull request closed without merging**
